Sprint 7 solution in file manager

IZarubitskiy opened the pull request:
ТЗ-7

IZarubitskiy committed:
ТЗ-7 начат

IZarubitskiy committed:
ТЗ-7 начат

IZarubitskiy committed:
ТЗ - 7 первая итерация

IZarubitskiy committed:
ТЗ - 7 первая итерация

IZarubitskiy committed:
ТЗ - 7 первая итерация

IZarubitskiy committed:
ТЗ - 7 первая итерация

IZarubitskiy committed:
ТЗ - 7 первая итерация

IZarubitskiy committed:
ТЗ - 7 первая итерация

IZarubitskiy committed:
ТЗ - 7 первая итерация

IZarubitskiy committed:
ТЗ - 7 первая итерация

PolnySkvorets reviewed (commented):
Напишу замечания в практикуме

IZarubitskiy committed:
ТЗ - 7 первая итерация

IZarubitskiy committed:
Исправление Ошибок

PolnySkvorets commented on src/FileBackedTaskManager.java:
```diff
@@ -53,18 +67,22 @@ public HashMap<Integer, SubTask> deleteSubTasks() {
     public SubTask addSubTask(SubTask newSubTask) {
+        super.addSubTask(newSubTask);
+        save();
         return super.addSubTask(newSubTask);
```
```suggestion
        SubTask subtask = super.addSubTask(newSubTask);
        save();
        return subtask;
```
У тебя выполняется лишнее добавление задачи
Выше тоже поправь.

PolnySkvorets commented on src/FileBackedTaskManager.java:
```diff
@@ -8,18 +8,32 @@
     private static final String HOME = System.getProperty("user.home");
+    File dir = new File("C:\\Users\\Igor\\java-kanban\\dbTaskManager");
 
+    File dbTaskManager = new File(dir, "dbTaskManager.csv");
```
Вот эти парметры должны задаваться из конструктора

IZarubitskiy replied:
не понимаю что значит должны задаваться из конструктора

PolnySkvorets reviewed (commented):
Оставил комментарии по правкам

IZarubitskiy committed:
Исправление Ошибок

IZarubitskiy committed:
Исправление Ошибок

IZarubitskiy committed:
3 вариант
Исправление ошибок
Добавление методов

PolnySkvorets commented on src/FileBackedTaskManager.java:
```diff
@@ -8,20 +8,22 @@
+    InMemoryHistoryManager inMemoryHistoryManager;
 
     public FileBackedTaskManager(File dbTaskManager, int id, int lastEpicId, HashMap<Integer, Task> singleTaskDesc, HashMap<Integer, Epic> epicTaskDesc, HashMap<Integer, SubTask> subTaskDesc, InMemoryHistoryManager inMemoryHistoryManager) {
         this.dbTaskManager = dbTaskManager;
```
```suggestion
    public FileBackedTaskManager(File dbTaskManager) {
        this. dbTaskManager = dbTaskManager;
        }
```

Конструктор по требованиям должен получиться примерно таким, все остальное: id, списки эпиков и остальных задач - инициализируется внутри менеджера. не совсем понятно зачем это     "File dir = new File("C:\\Users\\Igor\\java-kanban\\dbTaskManager");
    File dbTaskManager = new File("dbTaskManager.csv");"

PolnySkvorets commented on src/FileBackedTaskManager.java:
```diff
@@ -189,6 +190,77 @@ void reader() {
+        List<String> tasksList = new ArrayList<>();
+        Reader fileReader = null;
+        try {
+            Path dbTaskManagerDirectoryPath = Paths.get(HOME, "java-kanban", "dbTaskManager", "dbTaskManager.csv");
```
Это не используется

IZarubitskiy replied:
Я тут намудрил - плохо усвоил тему с путями и запихнул в общем все что знал. Исправил.

PolnySkvorets reviewed (commented):
Оставил замечание по конструктору

PolnySkvorets commented on src/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,216 @@
+    private HashMap<Integer, SubTask> subTaskDesc = new HashMap<>();
+    InMemoryHistoryManager inMemoryHistoryManager = new InMemoryHistoryManager();
+
+    public FileBackedTaskManager(File dbTaskManager, int id, int lastEpicId, HashMap<Integer, Task> singleTaskDesc, HashMap<Integer, Epic> epicTaskDesc, HashMap<Integer, SubTask> subTaskDesc, InMemoryHistoryManager inMemoryHistoryManager) {
```
По требованиям задачи этот конструктор должен принимать только file

PolnySkvorets commented on src/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,216 @@
+    public SubTask addSubTask(SubTask newSubTask) {
+        super.addSubTask(newSubTask);
+        save();
+        return super.addSubTask(newSubTask);
```
Здесь и в других методах нужно избавиться от дублирования
```suggestion
        SubTask subtask = super.addSubTask(newSubTask);
        save();
        return subtask;
```

IZarubitskiy replied:
странно - вроде вчера все методы поправил как было предложено. Исправлю

PolnySkvorets commented on src/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,216 @@
+import java.io.Reader;
+import java.io.File;
+
+public class FileBackedTaskManager extends InMemoryTaskManager  {
```
Здесь согласно требованиям не хватает статичного метода для загрузки списка задач из файла loadFromFile(File file)

IZarubitskiy committed:
4 вариант.
Исправил конструктор.
Поднял выше метод загрузки из файла

IZarubitskiy committed:
4 вариант(2)
Исправил конструктор.
Поднял выше метод загрузки из файла

Теперь принимает файл и работает уже только с ним

IZarubitskiy committed:
4 вариант(2)
Исправил конструктор.
Поднял выше метод загрузки из файла

Теперь принимает файл и работает уже только с ним

PolnySkvorets commented on src/FileBackedTaskManager.java:
```diff
@@ -1,152 +1,124 @@
-    File dir = new File("C:\\Users\\Igor\\java-kanban\\dbTaskManager");
 
-    File dbTaskManager = new File(dir, "dbTaskManager.csv");
+    File dbTaskManager = new File("dbTaskManager.csv");
```
Вот это лишнее, его не нужно вытри инициализировать. Поправь к следующей задаче.

PolnySkvorets commented on src/FileBackedTaskManager.java:
```diff
@@ -1,152 +1,124 @@
+            while (br.ready()) {
+                String line = br.readLine();
+                tasksList.add(line);
+            }
+            reader.close();
+            br.close();
 
-    @Override
-    public SubTask addSubTask(SubTask newSubTask) {
-        super.addSubTask(newSubTask);
-        save();
-        return super.addSubTask(newSubTask);
-    }
+        } catch (IOException e) {
+            e.printStackTrace();
+        } finally {
+            try {
+                fileReader.close();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            } catch (NullPointerException exp) { // ловим исключение NullPointerException
+                System.out.println("Ошибка: передан неинициализированный объект!");
+            }
+        }
```
чтобы не делать ещё и вложенную обработку исключений, ты можешь использовать try catch с закрываемыми ресурсами. Почитай здесь https://javarush.com/quests/lectures/questsyntaxpro.level15.lecture00

PolnySkvorets reviewed (approved):
Принято, но обрати внимание на замечания
